### ISSUE TITLE
Bundles ad-hoc

### ISSUE THREAD

**Issue by KarinG** [eileenmcnaughton/civicrm_entity#285]
Issues to be addressed:

**1. [this will no longer be an issue after 2] - Notices -> activity "add" form:** [KarinG: I have not been able to reproduce these but will try again and circle back here]
Notice: Undefined index: activity_type_id in Drupal\civicrm_entity\Entity\CivicrmEntity::preCreate() (line 87 of /var/www/web/demo8.skvare.com/web/modules/contrib/civicrm_entity/src/Entity/CivicrmEntity.php)

Notice: Undefined index: in Drupal\civicrm_entity\Entity\CivicrmEntity::preCreate() (line 90 of /var/www/web/demo8.skvare.com/web/modules/contrib/civicrm_entity/src/Entity/CivicrmEntity.php)

Notice: Undefined index: in Drupal\civicrm_entity\Plugin\Field\BundleFieldItemList->setValue() (line 60 of /var/www/web/demo8.skvare.com/web/modules/contrib/civicrm_entity/src/Plugin/Field/BundleFieldItemList.php)

**2. Bundles are showing up in places where they don't really need to be showing and that is confusing.** Also some larger clients can easily have 500+ activity types -> which results in lots of pages here. Easiest solution may be to not display the bundles where they are not needed. @mglaman thought he had already done this. He'll check into this -> 

![image](https://user-images.githubusercontent.com/5340555/114079010-de624380-9866-11eb-9ed2-23514f656391.png)

**3. Layout Builder issue - Manage display**  [KarinG: I'll try and reproduce - we don't use Layout Builder but obviously don't want to break anything].

cc: @puresyntax71 @mglaman @genaskvare @jackrabbithanna

**Comment by KarinG** [comment]
ad 1 -> trying to reproduce the notices - I don't see any notices/log messages in the watchdog

+Add CiviCRM Activity -> Activity
![image](https://user-images.githubusercontent.com/5340555/114081285-9a247280-9869-11eb-945d-4d1f3f9d4f7c.png)

Activity Created -> 
![image](https://user-images.githubusercontent.com/5340555/114081392-bde7b880-9869-11eb-8933-d305f7d38ea5.png)

If these notices only appear on Activity Add for Bundle types then 2 (hide Bundles) will solve this issue as well.

**Comment by KarinG** [comment]
ad 3 -> Layout Builder -> Installing Layout builder -> and enabling it on
/admin/structure/civicrm-entity/civicrm-activity/display

![image](https://user-images.githubusercontent.com/5340555/114082462-ffc52e80-986a-11eb-80ba-79467bbc24ab.png)

Then hitting Manage layout -> URL points to -> /admin/structure/civicrm-entity/civicrm-activity/display/default/layout?bundle=civicrm_activity

which results in a Page not found.

![image](https://user-images.githubusercontent.com/5340555/114082602-31d69080-986b-11eb-91bb-b586bc9c62a3.png)

For entities that have no bundles - like e.g. Addresses -> this URL
/admin/structure/civicrm-entity/civicrm-address/display/default/layout
gets you to the Form Builder screen you expect:

![image](https://user-images.githubusercontent.com/5340555/114082881-88dc6580-986b-11eb-92e7-d23f13ff8879.png)

**Comment by jackrabbithanna** [comment]
wrt notices .. Maybe check your PHP log? 
Using the 8.x-3.x branch of the github repo, not our fork (which only has Jon's PR and one other minor unrelated thing anyway)

I replicate like this:
goto /civicrm-activity/add
You'll see a list of bundles which are activity types. 
I click on any of them, for example "Meeting" 
so I get to a url like civicrm-activity/add/meeting

I get three notices:
Notice: Undefined index: activity_type_id in /var/www/html/demo8.skvare.com/web/modules/contrib/civicrm_entity/src/Entity/CivicrmEntity.php on line 87
Notice: Undefined index: in /var/www/html/demo8.skvare.com/web/modules/contrib/civicrm_entity/src/Entity/CivicrmEntity.php on line 90
Notice: Undefined index: in /var/www/html/demo8.skvare.com/web/modules/contrib/civicrm_entity/src/Plugin/Field/BundleFieldItemList.php on line 60

Testiing with 5.33 esr

**Comment by KarinG** [comment]
Ah see screenshot -> what I did (anticipating 2. No need to show all bundles) -> 

Add CiviCRM Activity -> **Activity** (top option).

Do you get any notices when just adding an Activity - and then selecting the Activity Type (as it used to be prior to bundles)?

**Comment by jackrabbithanna** [comment]
Ahh .. nope, I don't see those when using the top option.

**Comment by KarinG** [comment]
Ok so if we execute 2 -> 1 is ✅ - right?

**Comment by jackrabbithanna** [comment]
Sure, I don't see a reason to have to choose the activity type from a huge list of links. 
Then this solves the questions about whether to modify the add/edit forms .. disabling the activity_type_id field or not
After your update, it would just act like it used to, but on save, it would update the bundle property or the entity would be regarding of having the right bundle corresponding to activity type .. that's great 
People can just update activity type and the bundle stuff works itself out in the background. 
Sounds great to me.

**Comment by puresyntax71** [comment]
Yeah, I went ahead and added the changes for item 2 here - https://github.com/eileenmcnaughton/civicrm_entity/pull/286.

**Comment by puresyntax71** [comment]
Yeah, changes for item 3 there as well.

**Comment by KarinG** [comment]
Thank you @puresyntax71 - I’ll pull in theses commits and will review/test. 

@mglaman can help bring any remaining layout builder issue across the finish line. He told me he has done a lot of work with layout builder to make it work with commerce. He is trying to find a block of time - things are pretty crazy ahead of DrupalCon next week!

**Comment by KarinG** [comment]
Ok -> checking -> on one of my live test sites:

git log - Merge pull request #282 from puresyntax71/custom-field-required
curl https://patch-diff.githubusercontent.com/raw/eileenmcnaughton/civicrm_entity/pull/286.diff --output 286.diff
git apply 286.diff
sudo -u www-data drush cr

**Re: 1 and 2.** 👍 + Add CiviCRM Activity button now leads directly to -> /civicrm-activity/add - as it used to - no more confusion. I've created a new activity this way -> still works fine and nothing in logs. I've also created a non-bundle -> /civicrm-contact/add and that all still works fine as well - and also no log issues. 

And Calendar bits are still 👍 -> recoloured one activity type -> 
![image](https://user-images.githubusercontent.com/5340555/114182326-bd99fc80-98ff-11eb-9f59-64373ff9407b.png)

Re: 3 -> need to make the kids' lunches first :-)

**Comment by KarinG** [comment]
@puresyntax71 -> confirming we can now use Layout Builder on Manage Display 🎉
With this PR we no longer get a Page not Found and I can drag/drop/configure:

![image](https://user-images.githubusercontent.com/5340555/114224675-9dcdfd00-992e-11eb-888e-a1ccbb32d608.png)

Ok - so issue re: "entity view doesn't seem to use the layout builder to display"
I indeed don't see Layout Builder use this configured display on civicrm-activity/2846

Checking contact (no bundles)-> What is all this? There may be more issues with CE & Layout Builder?
![image](https://user-images.githubusercontent.com/5340555/114225128-319fc900-992f-11eb-8108-2631af15ebe8.png)

I did get this configured:
![image](https://user-images.githubusercontent.com/5340555/114225240-55630f00-992f-11eb-8d38-cb8e13db4595.png)

But also for e.g. civicrm-contact/212 - it's not using my Layout Builder config to display.

Ok checking before this PR - using 8.x-3.0-beta2

**Comment by KarinG** [comment]
karins-MBP:contrib sysadmin$ curl -L https://ftp.drupal.org/files/projects/civicrm_entity-8.x-3.0-beta2.tar.gz --output civicrm_entity-8.x-3.0-beta2.tar.gz
tar -xvf civicrm_entity-8.x-3.0-beta2.tar.gz
enable in UI
layout builder
civicrm entity

Ok -> Enabled CiviCRM Entity -> Activity -> Enable Layout Builder on it's Manage Display -> 

a) this is a completely different site & with beta2 release -> and I also see strange/odd strings:

![image](https://user-images.githubusercontent.com/5340555/114228555-d4f2dd00-9933-11eb-8f75-629991246aaa.png)

But I do see the columns -> it's using the Layout Builder config on /view -> 
![image](https://user-images.githubusercontent.com/5340555/114228638-f9e75000-9933-11eb-8a23-25ed7b8a6233.png)

**Comment by KarinG** [comment]
I don't see a tag here in Github -> was going to try do a /compare
Manual listing: last commit on drupal.org is: 790ee78f60c29e0b7ba995328b3b1251fbd2eef3 - which I don't see on Github? I could be blind - time for coffee here! https://git.drupalcode.org/project/civicrm_entity/-/commit/790ee78f60c29e0b7ba995328b3b1251fbd2eef3

In any case summary is: 
- I'll park the odd string/text in a separate issue [that's unrelated]
- it looks like Entity view no longer using the Layout Builder configured layout is connected to Bundles. I'll ask Matt to specifically focus on that.

**Comment by jackrabbithanna** [comment]
Arnold got some possible fixes: https://github.com/eileenmcnaughton/civicrm_entity/pull/286

**Comment by KarinG** [comment]
Yes 👍 I ran all my testing above 👆 with Arnold's PR 286 included (except of course for the beta2 testing) and we're really close.

**Comment by KarinG** [comment]
Update - after more testing -> civicrm-contact/ entities ad-here to layout -> but 'bundle' entities civicrm-activity/ do not -> they don't even get the Layout tab 

![image](https://user-images.githubusercontent.com/5340555/114269339-c5bb7000-99c3-11eb-991f-d5ea88521db2.png)

Config is the same:
![image](https://user-images.githubusercontent.com/5340555/114269432-45493f00-99c4-11eb-9b4f-0567adaa202c.png)

**Comment by mglaman** [comment]
Sorry for being late to the party, it's been a busy past 15 days with trainings and DrupalCon prep. I'm reviewing things, now.

**Comment by mglaman** [comment]
So when I use #286 and try to add an entity, I get the error. The error explained in scenario one is _exactly_ why I had to add the `add_form` and `add_page` routes. The easiest solution was just following Drupalisms. The way we could work around it is forcing the `add_form` route to have a default bundle provided.

<img width="1529" alt="Screen Shot 2021-04-11 at 8 00 45 AM" src="https://user-images.githubusercontent.com/3698644/114305043-0981a880-9a9c-11eb-9863-d9ee994a477b.png">

The error comes from `Drupal\Core\Entity\EntityForm->getEntityFromRouteMatch`

```
      // If the entity has bundles, fetch it from the route match.
      $entity_type = $this->entityTypeManager->getDefinition($entity_type_id);
      if ($bundle_key = $entity_type->getKey('bundle')) {
        if (($bundle_entity_type_id = $entity_type->getBundleEntityType()) && $route_match->getRawParameter($bundle_entity_type_id)) {
          $values[$bundle_key] = $route_match->getParameter($bundle_entity_type_id)->id();
        }
        elseif ($route_match->getRawParameter($bundle_key)) {
          $values[$bundle_key] = $route_match->getParameter($bundle_key);
        }
```

**Comment by mglaman** [comment]
Ah, I already have a slight work around here:

```
   if ($entity_type->hasKey('bundle')) {
      $bundle_property = $entity_type->get('civicrm_bundle_property');
      /** @var \Drupal\civicrm_entity\CiviCrmApiInterface $civicrm_api */
      $civicrm_api = \Drupal::service('civicrm_entity.api');
      $options = $civicrm_api->getOptions($entity_type->get('civicrm_entity'), $bundle_property);

      if (isset($values[$entity_type->getKey('bundle')]) && $values[$entity_type->getKey('bundle')] === $entity_type->id()) {
        $raw_bundle_value = key($options);
      }
      else {
        $raw_bundle_value = $values[$bundle_property];
      }
```

If the bundle is the same as the entity type ID it picks a default. So we just need to alter the `add_form` route. The only "problem" is types will be defaulted to the first option.

**Comment by mglaman** [comment]
The following PR addresses items one and two: https://github.com/eileenmcnaughton/civicrm_entity/pull/289

I'll work on Layout Builder later today.

**Comment by mglaman** [comment]
I think I found a bug in Layout Builder

```
      // If the entity type has no bundles and it doesn't use {bundle} in its
      // admin path, use the entity type.
      if (strpos($path, '{bundle}') === FALSE) {
        if (!$entity_type->hasKey('bundle')) {
          $defaults['bundle'] = $entity_type_id;
        }
        else {
          $defaults['bundle_key'] = $entity_type->getBundleEntityType();
        }
      }
```

The bundle key is wrong, well, its null.

**Comment by mglaman** [comment]
Nevermind, not Drupal core. That's basically a flag for Field UI's route enhancer, which we work around.

**Comment by mglaman** [comment]
I do think I found the reason the `Layout` task does not render. We "hide" all of the bundle displays. But the local task is dynamically generated and uses the current bundle magically. That hidden display hasn't been configured.

<img width="733" alt="Screen Shot 2021-04-11 at 3 53 49 PM" src="https://user-images.githubusercontent.com/3698644/114320978-3e165400-9ade-11eb-89a2-b92cebe4c4bb.png">

**Comment by mglaman** [comment]
Okay, so for Overrides the Layout tab doesn't display due to: The default storage access check \Drupal\layout_builder\Plugin\SectionStorage\DefaultsSectionStorage::access.

Overrides also ensures the default access check. This loads the display from context. That context uses the entity's bundle.

<img width="675" alt="Screen Shot 2021-04-11 at 4 08 57 PM" src="https://user-images.githubusercontent.com/3698644/114321397-3f488080-9ae0-11eb-9181-ccce49588cc8.png">

**Comment by mglaman** [comment]
Looks like I already added code in civicrm_entity_entity_view_display_alter to handle the fact we can't swap the display easily. Just need to copy third party settings, it seems

**Comment by mglaman** [comment]
https://github.com/eileenmcnaughton/civicrm_entity/pull/289 is ready. It supports Layout Builder with per-entity custom layout overrides, as well.

**Comment by KarinG** [comment]
I just tested this -> working as advertised. Thank you @mglaman 👍

1. No log messages
2. No more bundles on add
3. And Layout Builder is working 🎉
![image](https://user-images.githubusercontent.com/5340555/114324777-37d9a500-9ae9-11eb-844d-73ec1b6dec01.png)

**Comment by KarinG** [comment]
@jackrabbithanna @puresyntax71 @genaskvare

Matt has addresses the issues that originated from the Bundle work back in January. I have tested #289 myself -> see final screenshot to show you that Layout Builder is working now. 

Over to you! You're still on track to release another beta prior to Wednesday! 🎉

**Comment by genaskvare** [comment]
Nice work everyone!

**Comment by puresyntax71** [comment]
Yeah, I went ahead and closed #286. #289 working on my end as well.

**Comment by mglaman** [comment]
Should this issue be closed since #289 is merged? Or should we track that remaining notice here (someone should copy the text over and I can work on it.)

**Comment by KarinG** [comment]
Closing this - as I just had a got at that notice on PHP7.4 and still not able to reproduce it. It needs its own Issue.
